source/opt/optimizer.cpp

```diff
@@ -540,26 +540,24 @@ bool Optimizer::Run(const uint32_t* original_binary,
   impl_->pass_manager.SetTargetEnv(impl_->target_env);
   auto status = impl_->pass_manager.Run(context.get());
 
-  bool binary_changed = false;
-  if (status == opt::Pass::Status::SuccessWithChange) {
-    binary_changed = true;
-  } else if (status == opt::Pass::Status::SuccessWithoutChange) {
+  if (status == opt::Pass::Status::Failure) {
+    return false;
+  }
+
+  optimized_binary->clear();
+  context->module()->ToBinary(optimized_binary, /* skip_nop = */ true);
+
+  if (status == opt::Pass::Status::SuccessWithoutChange) {
     if (optimized_binary->size() != original_binary_size ||
         (memcmp(optimized_binary->data(), original_binary,
                 original_binary_size) != 0)) {
-      binary_changed = true;
       Log(consumer(), SPV_MSG_WARNING, nullptr, {},
```

Comment:
This strikes me as odd.  If the optimizer reports that there was no change, but it did change, then that is a bug in the optimizer.  Why would the user need to be warned about that.  They can do nothing about it. Warning and error should be warning or error about something the user did.  Not bugs in the compiler.

I think we should change this to only do the comparison an assert, which is only enabled in a debug build.  Then for a release build, we only need to write to `optimized_binary` if the optimizer reported a change or `optimized_binary != original_binary`.

Reply:
Couldn't the user be providing their own optimization passes?

Reply:
They  could, but at that point they are now compiler developers.  The error is in their role as a compiler developer, not as a user.

This warning was added in #2487.  I should have picked up on it in the code review for that PR.

Reply:
Okay, migrated to an assert.